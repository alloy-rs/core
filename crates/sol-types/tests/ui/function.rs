use alloy_sol_types::sol;

sol! {
    function missingParens;
}

sol! {
    function missingSemi1()
}

sol! {
    function missingSemi2() external
}

sol! {
    function missingSemi3() returns (uint256)
}

// OK
sol! {
    function semiNotBrace1() {}
}

// OK
sol! {
    function semiNotBrace2() external {}
}

// OK
sol! {
    function semiNotBrace3() returns (uint256) {}
}

sol! {
    function singleComma(,);
}

// OK
sol! {
    function trailingComma1(bytes,);
    function trailingComma2(bytes a,);
    function trailingComma3(bytes memory a,);
}

sol! {
    function badReturn1() returns;
}

sol! {
    function badReturn2() returns();
}

// OK
sol! {
    function a() private;
    function b() internal;
    function c() public;
    function d() external;

    function e() pure;
    function f() view;
    function g() constant;
    function h() payable;

    function i() virtual;
    function j() immutable;

    function k() override(Interface.k);
    function l() myModifier("a", 0);

    function m() external view returns (uint256);
    function n() public pure returns (uint256,);
}

// OK
sol! {
    function overloaded();
    function overloaded(uint256);
    function overloaded(uint256, address);
    function overloaded(address);
    function overloaded(address, string);
}

sol! {
    function overloadTaken();
    function overloadTaken(uint256);

    function overloadTaken_0();
    function overloadTaken_1();
    function overloadTaken_2();
}

sol! {
    function sameFnOverload();
    function sameFnOverload();
}

sol! {
    function sameFnTysOverload1(uint256[] memory a);
    function sameFnTysOverload1(uint256[] storage b);
}

sol! {
    function sameFnTysOverload2(string memory, string storage);
    function sameFnTysOverload2(string storage b, string calldata);
}

fn main() {}
