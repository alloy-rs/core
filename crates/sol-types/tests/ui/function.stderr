error: expected parentheses
 --> tests/ui/function.rs:4:27
  |
4 |     function missingParens;
  |                           ^

error: unexpected end of input, expected curly braces or `;`
 --> tests/ui/function.rs:7:1
  |
7 | / sol! {
8 | |     function missingSemi1()
9 | | }
  | |_^
  |
  = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of input, expected curly braces or `;`
  --> tests/ui/function.rs:11:1
   |
11 | / sol! {
12 | |     function missingSemi2() external
13 | | }
   | |_^
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of input, expected curly braces or `;`
  --> tests/ui/function.rs:15:1
   |
15 | / sol! {
16 | |     function missingSemi3() returns (uint256)
17 | | }
   | |_^
   |
   = note: this error originates in the macro `sol` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected a Solidity type: `address`, `bool`, `string`, `bytesN`, `intN`, `uintN`, `tuple`, `function`, `mapping`, or a custom type name
  --> tests/ui/function.rs:35:26
   |
35 |     function singleComma(,);
   |                          ^

error: expected parentheses
  --> tests/ui/function.rs:46:34
   |
46 |     function badReturn1() returns;
   |                                  ^

error: expected at least one return type
  --> tests/ui/function.rs:50:34
   |
50 |     function badReturn2() returns();
   |                                  ^^

error: function `overloadTaken` is overloaded, but the generated name `overloadTaken_0` is already in use
  --> tests/ui/function.rs:85:14
   |
85 |     function overloadTaken();
   |              ^^^^^^^^^^^^^

error: other declaration is here
  --> tests/ui/function.rs:88:14
   |
88 |     function overloadTaken_0();
   |              ^^^^^^^^^^^^^^^

error: function `overloadTaken` is overloaded, but the generated name `overloadTaken_1` is already in use
  --> tests/ui/function.rs:86:14
   |
86 |     function overloadTaken(uint256);
   |              ^^^^^^^^^^^^^

error: other declaration is here
  --> tests/ui/function.rs:89:14
   |
89 |     function overloadTaken_1();
   |              ^^^^^^^^^^^^^^^

error: function with same name and parameter types defined twice
  --> tests/ui/function.rs:94:14
   |
94 |     function sameFnOverload();
   |              ^^^^^^^^^^^^^^

error: other declaration is here
  --> tests/ui/function.rs:95:14
   |
95 |     function sameFnOverload();
   |              ^^^^^^^^^^^^^^

error: function with same name and parameter types defined twice
  --> tests/ui/function.rs:99:14
   |
99 |     function sameFnTysOverload1(uint256[] memory a);
   |              ^^^^^^^^^^^^^^^^^^

error: other declaration is here
   --> tests/ui/function.rs:100:14
    |
100 |     function sameFnTysOverload1(uint256[] storage b);
    |              ^^^^^^^^^^^^^^^^^^

error: function with same name and parameter types defined twice
   --> tests/ui/function.rs:104:14
    |
104 |     function sameFnTysOverload2(string memory, string storage);
    |              ^^^^^^^^^^^^^^^^^^

error: other declaration is here
   --> tests/ui/function.rs:105:14
    |
105 |     function sameFnTysOverload2(string storage b, string calldata);
    |              ^^^^^^^^^^^^^^^^^^
